rclpy/rclpy/node.py

```diff
@@ -168,29 +169,34 @@ def __init__(
             to get and set logger levels of this node. Otherwise, logger levels are only managed
             locally. That is, logger levels cannot be changed remotely.
         """
-        self.__handle = None
         self._context = get_default_context() if context is None else context
-        self._parameters: Dict[str, Parameter] = {}
-        self._publishers: List[Publisher] = []
-        self._subscriptions: List[Subscription] = []
-        self._clients: List[Client] = []
-        self._services: List[Service] = []
+        self._parameters: Dict[str, Parameter[Any]] = {}
+        self._publishers: List[Publisher[Any]] = []
+        self._subscriptions: List[Subscription[Any]] = []
+        self._clients: List[Client[Any, Any, Any]] = []
+        self._services: List[Service[Any, Any, Any]] = []
         self._timers: List[Timer] = []
         self._guards: List[GuardCondition] = []
         self.__waitables: List[Waitable[Any]] = []
         self._default_callback_group = MutuallyExclusiveCallbackGroup()
-        self._pre_set_parameters_callbacks: List[Callable[[List[Parameter]], List[Parameter]]] = []
+        self._pre_set_parameters_callbacks: List[Callable[[List[Parameter[Any]]],
+                                                          List[Parameter[Any]]]] = []
         self._on_set_parameters_callbacks: \
-            List[Callable[[List[Parameter]], SetParametersResult]] = []
-        self._post_set_parameters_callbacks: List[Callable[[List[Parameter]], None]] = []
+            List[Callable[[List[Parameter[Any]]], SetParametersResult]] = []
+        self._post_set_parameters_callbacks: List[Callable[[List[Parameter[Any]]], None]] = []
         self._rate_group = ReentrantCallbackGroup()
         self._allow_undeclared_parameters = allow_undeclared_parameters
-        self._parameter_overrides: Dict[str, Parameter] = {}
+        self._parameter_overrides: Dict[str, Parameter[Any]] = {}
         self._descriptors: Dict[str, ParameterDescriptor] = {}
 
         namespace = namespace or ''
+
         if not self._context.ok():
             raise NotInitializedException('cannot create node')
+
+        if self._context.handle is None:
+            raise RuntimeError('Context must be initialized before a node can be created')
+
```

Comment:
i wonder if we need to have this check for the `Context.handle` because above call `self._context.ok()` already verifies that context is valid. IMO we can remove this change for now, and not directly related to the objective for this PR? what do you think?

Reply:
I agree that `self._context.ok()` does the same check however in Python's type system there is no way to have a function give information about other attributes.


Would the following be cleaner in your opinion?
```python
if self._context.handle is None or not self._context.ok():
    raise NotInitializedException('cannot create node')
 ```

Reply:
makes sense to me! thanks for the suggestion